File: leetcode/editor/en/[68]Text Justification.java

//leetcode submit region begin(Prohibit modification and deletion)
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> list = new ArrayList<>();
        int n = words.length;
        int i = 0;
        while (i < n) {
            int j = i + 1;
            int lineLen = words[i].length();
            while (j < n && lineLen + words[j].length() + (j - i - 1) < maxWidth) {
                lineLen += words[j].length();
                j++;
            }
            int diff = maxWidth - lineLen;
            int numWords = j - i;
            StringBuilder sb = new StringBuilder();
            sb.append(words[i]);
            if (numWords == 1 || j >= n) {
                for (int k = i + 1; k < j; k++) {
                    sb.append(" ");
                    sb.append(words[k]);
                }
                sb.append(" ".repeat(maxWidth - sb.length()));
            } else {
                int spaces = diff / (numWords - 1);
                int extraSpaces = diff % (numWords - 1);

                for (int k = i + 1; k < j; k++) {
                    int spacesToApply = spaces + (extraSpaces-- > 0 ? 1 : 0);
                    sb.append(" ".repeat(spacesToApply));
                    sb.append(words[k]);
                }
            }
            list.add(sb.toString());
            i = j;
        }
        return list;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/en/[68]Text Justification.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {string[]} words
 * @param {number} maxWidth
 * @return {string[]}
 */
var fullJustify = function (words, maxWidth) {
    let res = [];
    let line = [];
    let lineLength = 0;
    let i = 0;
    while (i < words.length) {
        if (lineLength + words[i].length <= maxWidth) {
            line.push(words[i]);
            lineLength += words[i].length + 1;
            i++;
        } else {
            res.push(line);
            line = [];
            lineLength = 0;
        }
    }
    if (line.length > 0) {
        res.push(line);
    }

    for (let i = 0; i < res.length - 1; i++) {
        let line = res[i];
        if (line.length === 1) {
            res[i] = line[0] + " ".repeat(maxWidth - line[0].length);
        } else {
            let space = maxWidth - line.reduce((sum, word) => sum + word.length, 0);
            let spacePerWord = Math.floor(space / (line.length - 1));
            let extraSpace = space % (line.length - 1);
            let lineStr = "";
            for (let j = 0; j < line.length - 1; j++) {
                lineStr += line[j] + " ".repeat(spacePerWord + (j < extraSpace ? 1 : 0));
            }
            lineStr += line[line.length - 1];
            res[i] = lineStr;
        }
    }
    let lastLine = res[res.length - 1];
    res[res.length - 1] = lastLine.join(" ") + " ".repeat(maxWidth - lastLine.reduce((sum, word) => sum + word.length, 0) - lastLine.length + 1);
    return res;
};
//leetcode submit region end(Prohibit modification and deletion)
